docs/design/quota.md

```diff
@@ -0,0 +1,121 @@
+* Set a quota of object bucket claim, which limits the max number of objects per bucket (MaxObjects).
+* Set a quota of object bucket claim, which limits the max total size per bucket (MaxSize).
+* Remove quota configuration from the bucket. **?**
+* Update Quota’s on exiting buckets (updating the OBC only).
+* Update BucketClass quota, will be effective in new buckets using this BucketClass.
+* Add to NooBaa CLI new options for creating object bucket claims while providing one of the 2 types of quota.
+* Metrics and Alerts - upon calculation statistics based:
+    * Will provide the admin the quota status.
+    * Alert the admin when the number of objects / total size of objects reaches 80% and 100% of the quota. **# in the code 90% and 100% threashold?**
+
+## Out of scope
+* Quota on Namespace OBCs.
+* Quota on Cache OBCs.
+
+## Technical Details
+* Versioning Enabled Bucket
+    * Each version counts as a different object towards num objects quota and its size towards size quota.
+    * Delete marker is counted as a new object towards num objects quota. This means deletion of a non specific version can fail if num objects quota exceeded.
+    * Delete for a specific version would work, since its not creating a delete marker.
+* Non Versioned Bucket
+    * Delete should always succeed
+* Multipart Uploads
+    * If multipart upload init succeeded, the upload can be completed. Quota will be checked on init to avoid an increasing number of uncompleted uploads taking up space.
+### Limitations
+* Non strict quota - quota calculation is based on statistics aggregation. 
+    * Statistics calculation cycle is executed every few minutes (2m) - will only limit the user if he exceeded the quota after the next statistics calculation cycle will end.
+
+
+## Solution
+
+### **noobaa-core component**
+  
+Noobaa-core component uses rpc server with rich API that allows to communicate with client based on JSON schema.  
+buckets manipulations is described in bucket_api schema (bucket_api.js)
+* update_buckets - update buckets with quota config based on provided buckets list.
+* update_bucket - update single bucket with quota configuration. internally uses update_buckets.
+* read_bucket/read_bucket_sdk_info - return bucket_info definition which includes bucket quota configuration. 
+
+quota config schema:
+
+```
+quota_config: {
+    type: 'object',
+    properties: {
+        size: {
+            type: 'object',
+            required: ['value', 'unit'],
+            properties: {
+                value: {
+                    type: 'integer'
+                },
+                unit: {
+                    type: 'string',
+                    enum: ['GIGABYTE', 'TERABYTE', 'PETABYTE']
+                },
+            },
+        },
+        amount: {
+            type: 'object',
+            required: ['value'],
+            properties: {
+                value: {
+                    type: 'integer'
```

Comment:
I'm not sure if we need here `integer` or `bigint`. I think integer is enough. @nimrod-becker WDYT?

Reply:
I think Integer is enough, MAX_SAFE_INTEGER even in bytes is still high enough and I Don't expect we'd overflow it